components/application-gateway/internal/authorization/certificategenstrategy.go

```diff
@@ -20,7 +19,7 @@ func newCertificateGenStrategy(certificate, privateKey []byte) certificateGenStr
 	}
 }
 
-func (b certificateGenStrategy) AddAuthorization(r *http.Request, proxy *httputil.ReverseProxy) apperrors.AppError {
+func (b certificateGenStrategy) AddAuthorization(r *http.Request, setter TransportSetter) apperrors.AppError {
```

Comment:
Consider renaming `setter` to something more accurate

Reply:
As we know, naming things is one of two things really hard in computer science. Please suggest a better name ;)

Reply:
Oh yeah, naming is the most difficult stuff! But I'm going to ask other question. What is the idea behind introducing TransportSender interface? I can't see an issue with setting transport directly on proxy.  After this change AddAuthorization method is not aware how the request will be made (which is neither good or bad in my opinion).

Reply:
@akgalwas AddAuthorization is now used in two different scenarios:
- regular calls to upstream service;
- requests sent to the service endpoint that exposes CSRF tokens (secured with the same auth method as regular calls).

If case of certificateGen strategy, we would need to create another instance of proxy to add authorization to requests. @Tomasz-Smelcerz-SAP and I discussed the topic with @Szymongib, he should know all the details.

Reply:
@akgalwas about reasons for the TransportSetter function
Take a look at how AddAuthorization was implemented before: It took a reverseProxy as an argument, and then, only in one particular implementation (certificategenstrategy.go), it was setting a Transport object on the Proxy. Later on the _Client_ used by the Proxy can make API requests using this Transport.
In this PR we have another _Client_: a client to fetch CSRF tokens. This client is not related to proxy, but it has to use the same authorization method as the Proxy Client (it is an Acceptance Criteria for the task).
It turns out that we have to re-use the same Authorization strategy for another client, but this time we don't have any "Proxy" object to pass to "AddAuthorization"...
We've had two choices:
- Provide a dummy "Proxy" object  For CSRF client just to satisfy AddAuthorization signature and setup Transport on CSRF Client properly
- Refactor AddAuthorization interface to be more specific. AddAuthorization actually never required access to the entire _Proxy_ object, all it was doing was setting a Transport. Applying the "Principle of Least Responsibility" here leads us to the interface that allows just that: setting a transport

We have chosen the second option, after consulting this choice with @Szymongib 
I hope that explains our design choices. The name is still bad, though :)